### ISSUE TITLE
[Discussion] How to allow for more dynamic prompt_embed scaling/weighting/fusion?

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#4778]
We have a couple of issues and requests for the community that ask for the possibility to **dynamically** change certain knobs of Stable Diffusion that are applied at **every denoising step**. 

- 1. **Prompt Fusion**. as stated [here](https://github.com/huggingface/diffusers/issues/4496). To implement prompt fusion in a general way we need to give the user the possibility to define some kind of "prompt" scheduler where every denoising timestep can receive a different `prompt_embeds` and `negative_prompt_embeds`. 

=> A very obvious way to allow for this would be to allow passing a list of list of prompts and list of list of `prompt_embeddings`

- 2. **Dynamic prompt weighting**. A1111 and InvokeAI both have functionalities that allow to weight the prompt embeddings differently at each timestep. InvokeAI has this implemented in `compel` via a `conditioning_scheduler` see here: https://github.com/damian0815/compel/blob/d15e883bbbfae5b3fbd8d60065aa330c99a662b4/src/compel/compel.py#L93 
Such a scheduler could for example allow the user to not just define a unique `prompt_embedding` condition (e.g. putting more word on a certain word), but also allowing to dynamically change that condition during the course of denoising.
This is also asked by SD.Next (cc @vladmandic).

=> Here we have a couple of options, the simplest is probably to just allow passing a list of `prompt_embeddings` assuming that the user just takes care of the prompt weighting themselves. We could then also nicely integrate this with `compel`.

- 3. **Dynamic `guidance_scale` / `cfg` weighting**. Many people have found that a `cfg` scheduling works really well for `SDXL`. It's related to 2. as it's also a knob to tweak text embeddings weights over the course of inference but it's much more global where as 2. is can be more condition specific. This is also related to https://github.com/huggingface/diffusers/pull/4569#issuecomment-1678667625 which proposes dynamic scaling.

=> Here we could solve this by allowing the user to provide a list of `guidance_scales`. In addition we could maybe introduce something like `guidance_scaling_type="static/dynamic" to allow for #4569 

**Overall**:

=> It's not too difficult to make these features work, but it'll require some very good docs about `prompt_embeds` and `negative_prompt_embeds`. We also have to think about edge cases like SDXL which has two text encoders. We also have to think about how this can be applied to other models such as Kandinsky, IF.

Curios to hear your thoughts here. Also would love to discuss a design proposal of how we can better support things in a coherent, library-wide design @sayakpaul @williamberman @yiyixuxu @DN6

**Comment by vladmandic** [comment]
question is how far do diffusers want to go into auto-managing all border-line cases vs potentially exposing some inner workings so any such work can be done externally?

my $0.02 - trying to cater to each and every case is not the best approach as there will always be "one more". plus it limits ability to experiment (and potentially come up with a new cool ideas/workflows).

for example, looking at `StableDiffusionPipeline`, `__call__` implementation - instead of assigning *args* to local variables, it could be a dict or class (for example `instead of `prompt` variable, there could be `args.prompt`) - and then entire args object can be passed when using `callback` method.

that way any internal value can be changes by external code per-step.
e.g., app wants to change prompt_embeds between step 12 and 13? sure. change guidance scale? sure.
it opens a lot more options than trying to come up with design for each individual ask.

for example, i have an ask to perform early-exit in case inference is going in undesired direction. sure, i can modify timesteps directly. are such things inherently safe? no. but its really flexible and powerful.

**Comment by williamberman** [comment]
yes, I think passing as a list of lists makes the most sense if we want this to be integrated into the main pipelines class. Probably need to be careful since a list vs a list of lists being two different arguments which have different meanings can be a bit confusing

**Comment by williamberman** [comment]
@vladmandic understand where you're coming from but I've always had a real aversion to passing large numbers of arguments around in dicts in python. It makes it really hard to read the code and remember which arguments are inside the dicts. 

I think an alternative to provide a strong degree of flexibility would be to have a well documented minimal sampling loop which can then be manually tweaked by the user.

i.e. we flesh an example like this out

```py
pipeline = Pipeline()

prompt = "foo bar"
prompt_embeddings = pipeline.encode_text(prompt)

timesteps = pipeline.scheduler.timesteps

latents = pipeline.vae.encode(<image>)

for timestep in timesteps:
    latents = pipeline.unet(latents, timestep)
    latents = pipeline.scheduler.step(latents)
```

and then the user can make manual modifications to the sampling loop if needed

**Comment by vladmandic** [comment]
That also works, but it basically reimplemented call method.

Re:dict - I agree python dict is not ideal - a simple class to hold args is more elegant.

**Comment by alexblattner** [comment]
I have made something that solves the lack of flexibility you have for your pipelines in general. The reason you even rose this issue is because without this, you would need to create a pipeline for all other features with prompt fusion which is a huge headache.

Take a look at my solution: https://github.com/alexblattner/RubberDiffusers

This makes it so every code is easily reproducible and doesn't require a new pipeline for every feature variation. I made prompt fusion work on it too. 
@patrickvonplaten  @vladmandic  @williamberman 

I would appreciate feedback on this if you don't mind.

**Comment by patrickvonplaten** [comment]
@yiyixuxu would you maybe like to give this issue a try? I think it'll be very important to make sure diffusers stays flexible going forward :-)

**Comment by vladmandic** [comment]
i think this should now be resolved.

**Comment by yiyixuxu** [comment]
you can use callback function to achieve this now
check out this PR https://github.com/huggingface/diffusers/pull/5427
